sharktank/sharktank/examples/pipeline/export_ppffn_net.py

```diff
@@ -77,57 +92,55 @@ def create_theta(dim: int, shard_count: int, num_layers: int, save_path):
 
 
 class PPFFN(ThetaLayer):
-    block_to_pipeline: tuple[int, ...]
-    pipeline_to_devices: tuple[list[int], ...]
-
     def __init__(
         self,
         theta,
-        block_to_pipeline: tuple[int, ...],
-        pipeline_to_devices: tuple[list[int], ...],
+        block_to_pipeline_stage: list[int],
+        pipeline_stage_to_devices: list[list[int]],
     ):
         super().__init__(theta)
-        self.block_to_pipeline = block_to_pipeline
-        self.pipeline_to_devices = pipeline_to_devices
-
-    def _inter_layer_callback(self, x: ShardedTensor, curr_block: int):
-        if self.block_to_pipeline is None:
-            return x
-
-        if curr_block >= len(self.block_to_pipeline) - 1:
-            return x
-
-        curr_pipeline = self.block_to_pipeline[curr_block]
-        next_pipeline = self.block_to_pipeline[curr_block + 1]
-
-        curr_devices = self.pipeline_to_devices[curr_pipeline]
-        next_devices = self.pipeline_to_devices[next_pipeline]
-
-        if all(d_curr == d_next for d_curr, d_next in zip(curr_devices, next_devices)):
-            return x
+        self.block_to_pipeline_stage = block_to_pipeline_stage
+        self.pipeline_stage_to_devices = pipeline_stage_to_devices
 
-        shards = ShardedTensor.move_shards_to_new_devices(
-            x.shards, old_devices=curr_devices, new_devices=next_devices
+        self.blocks = torch.nn.ModuleList(
+            LinearLayer(theta(f"blk.{block_idx}.ffn"))
+            for block_idx in range(len(block_to_pipeline_stage))
         )
-        return x.clone(ts=shards, devices=next_devices)
 
     def forward(self, x: torch.Tensor):
-        num_blocks = len(self.block_to_pipeline)
-        shard_count = self.theta.tensor("blk.0.ffn.weight").shard_count
-
+        # Note: In export_paged_llm, this replicate happens outside of forward().
+        #       Done here in this example for simplicity.
         x = ReplicatedTensor(
-            ts=x, shard_count=shard_count, devices=self.pipeline_to_devices[0]
+            ts=x,
+            shard_count=len(self.pipeline_stage_to_devices[0]),
+            devices=self.pipeline_stage_to_devices[0],
         )
-        for block in range(num_blocks):
-            weight = self.theta.tensor(f"blk.{block}.ffn.weight")
-            x = ops.replicate(ops.linear(x, weight), shard_count)
-            x = self._inter_layer_callback(x, block)
+        for block_idx, block in enumerate(self.blocks):
+            x = block(x)
+            x = transfer_between_blocks_if_needed(
```

Comment:
In the follow up - lets do a "transfer_between_blocks_if_needed" where you pass in the `theta` for the block. Then you just process determine the devices to transfer to based on the placement of those tensors. This way we no longer need to compute / pass a staging map.